doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,530 @@
+        }
+        op: REPLACE
+    }
+    response {
+        path {
+            origin: "sonic_yang"
+            elem {name: "CONFIG_DB"} elem {name: "sonic-device_neighbor:sonic-device_neighbor"} elem {name: "sonic-device_neighbor:DEVICE_NEIGHBOR"} elem {name: "DEVICE_NEIGHBOR_LIST" key {key: "name" value: "Ethernet8"}} elem {name: "port"}
+        }
+        op: REPLACE
+    }
+
+#### 1.2.1.4 Incremental Configurations
+
+For incremental configurations, SetRequest can use all kinds of operations, for example, 'replace' operations, 'update' operations and 'delete' operations. And generic_config_updater can guarantee the outcome is correct.
+
+Generic_config_updater is used for incrementally update configurations, and it has two related API:
+1. replace, input format is ConfigDB schema.
+2. apply-patch, input format is JsonPatch.
+
+SetRequest syntax is not the same as JsonPatch. For example, 'replace' operation can be used to delete a node and update a node, please refer to gnmi specification.
+
+We can translate most of the SetRequest operations directly:
+
+### Table 2: Mapping Operations
+| SetRequest Operation | Description | JsonPatch Operation |
+|  ----  | ----  | ---- |
+| Delete | Where a path is contained within the delete field of the SetRequest message, it should be removed from the target's data tree. | Remove |
+| Update | If a particular path-value is specified in the client-supplied data, it is replaced with the client-specified value. <br>If the path specified does not exist, the target MUST create the data tree element and populate it with the data in the Update message. | Add |
+| Replace (delete) | If a particular path-value is NOT specified in the client-supplied data and the path does not have a specified default value in the corresponding schema, it should be deleted. | Remove |
+| Replace (update) | If a particular path-value is specified in the client-supplied data, it is replaced with the client-specified value.<br>If the path specified does not exist, the target MUST create the data tree element and populate it with the data in the Update message. | Add |
+
+Below table provides an example to translate from SetRequest to JsonPatch.
+
+### Table 3: Translate Example
+| SetRequest Format | JsonPatch Format |
+|  ----  | ----  |
+| replace {<br>&ensp;path {<br>&ensp;&ensp;elem {name: "CONFIG_DB"} elem {name: "DEVICE_NEIGHBOR"} elem {name: "Ethernet96"}<br>&ensp;}<br>}<br>replace {<br>&ensp;path {<br>&ensp;&ensp;elem {name: "CONFIG_DB"} elem {name: "DEVICE_NEIGHBOR"} elem {name: "Ethernet8"} elem {name: "port"}<br>&ensp;}<br>&ensp;val { json_ietf_val: "eth1" }<br>} | [<br>&ensp;{<br>&ensp;&ensp;"op": "remove", "path": "/DEVICE_NEIGHBOR/Ethernet96"}, <br>&ensp;&ensp;{ "op": "add", "path": "/DEVICE_NEIGHBOR/Ethernet8/port", "value": "eth1"}<br>] |
+
+#### 1.2.1.5 Full Configurations
```

Comment:
Hows the simultaneous configuration from CLI and GNMI  protected/serialized ?

Reply:
Totally agree. Simultaneous configuration from CLI and GNMI should be protected.
We will update CLI implementation, and it's not covered in this design.

Reply:
Thank you, please enhance the HLD to add appropriate protection.

Reply:
@bandaru-viswanath this is a great question! Currently the HLD of gNMI server does not protect against confliction with CLI, similar to one CLI does not protect against confliction with another CLI today.

In the future vision, there should be a centralized service inside SONiC responsible for all kinds of configuration. The requests could come from CLI/gNMI/RestAPI or SONiC applications. And the request should be serialized if there is confliction. The gNMI server or a sub service under gNMI server may be a candidate for such centralized service.